pkg/nodeidentity/scaleway/identify.go

```diff
@@ -0,0 +1,155 @@
+func New(CacheNodeidentityInfo bool) (nodeidentity.Identifier, error) {
+	scwClient, err := scw.NewClient(
+		scw.WithUserAgent("kubernetes-kops/"+kopsv.Version),
+		scw.WithEnv(),
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	return &nodeIdentifier{
+		client:       scwClient,
+		cache:        expirationcache.NewTTLStore(stringKeyFunc, cacheTTL),
+		cacheEnabled: CacheNodeidentityInfo,
+	}, nil
+}
+
+// IdentifyNode queries Scaleway for the node identify information
+func (i *nodeIdentifier) IdentifyNode(ctx context.Context, node *corev1.Node) (*nodeidentity.Info, error) {
+	providerID := node.Spec.ProviderID
+	if providerID == "" {
+		return nil, fmt.Errorf("providerID not set for node %s", node.Name)
+	}
+	if !strings.HasPrefix(providerID, "scaleway://") {
+		return nil, fmt.Errorf("providerID %q not recognized for node %s", providerID, node.Name)
+	}
+
+	serverID := strings.TrimPrefix(providerID, "scaleway://")
+
+	// If caching is enabled try pulling nodeidentity.Info from cache before doing a Scaleway API call.
+	if i.cacheEnabled {
+		obj, exists, err := i.cache.GetByKey(serverID)
+		if err != nil {
+			klog.Warningf("Nodeidentity info cache lookup failure: %v", err)
+		}
+		if exists {
+			return obj.(*nodeidentity.Info), nil
+		}
+	}
+
+	server, err := i.getServer(ctx, serverID)
+	if err != nil {
+		return nil, err
+	}
+
+	if server.State != instance.ServerStateRunning && server.State != instance.ServerStateStarting {
+		return nil, fmt.Errorf("found server %s (%s) with unexpected state: %q", server.Name, server.ID, server.State)
+	}
+
+	labels := map[string]string{}
+	for _, tag := range server.Tags {
+		if strings.HasPrefix(tag, scaleway.TagNameRolePrefix) {
+			role := strings.TrimPrefix(tag, scaleway.TagNameRolePrefix+"=")
+			switch kops.InstanceGroupRole(role) {
+			case kops.InstanceGroupRoleControlPlane:
+				labels[nodelabels.RoleLabelControlPlane20] = ""
+			case kops.InstanceGroupRoleNode:
+				labels[nodelabels.RoleLabelNode16] = ""
+			case kops.InstanceGroupRoleAPIServer:
+				labels[nodelabels.RoleLabelAPIServer16] = ""
+			default:
+				klog.Warningf("Unknown node role %q for server %s(%d)", role, server.Name, server.ID)
+			}
+		}
+	}
+
+	info := &nodeidentity.Info{
+		InstanceID: serverID,
+		Labels:     labels,
+	}
+
+	// If caching is enabled add the nodeidentity.Info to cache.
+	if i.cacheEnabled {
+		err = i.cache.Add(info)
+		if err != nil {
+			klog.Warningf("Failed to add node identity info to cache: %v", err)
+		}
+	}
+
+	return info, nil
+}
+
+// stringKeyFunc is a string as cache key function
+func stringKeyFunc(obj interface{}) (string, error) {
+	key := obj.(*nodeidentity.Info).InstanceID
+	return key, nil
+}
+
+// getServer queries Scaleway for the server with the specified ID, returning an error if not found
+func (i *nodeIdentifier) getServer(ctx context.Context, id string) (*instance.Server, error) {
+	api := instance.NewAPI(i.client)
+	zone := os.Getenv("SCW_DEFAULT_ZONE")
```

Comment:
Would it be possible to get the zone through configuration instead of the environment? Not a blocking issue.

Reply:
I will eventually try to get rid of the environment variables as you suggested in your review of  #14735, for now I just removed it from here.